File: githand/core/discover.py
"""Directory scanning — discover git repos under a path."""

from __future__ import annotations

from pathlib import Path

from githand.core.models import RepoRecord
from githand.core.utils import is_git_repo


def scan_directory(
    base_path: Path,
    recursive: bool = False,
    auto_group: bool = False,
) -> list[RepoRecord]:
    """Scan a directory for git repositories.

    Args:
        base_path: root directory to scan
        recursive: scan subdirectories recursively
        auto_group: use subdirectory name as group tag

    Returns:
        list of RepoRecord for discovered repos
    """
    records: list[RepoRecord] = []
    seen: set[str] = set()

    _walk(base_path, base_path, records, seen, recursive=recursive, auto_group=auto_group)
    return records


def _walk(
    current: Path,
    base: Path,
    records: list[RepoRecord],
    seen: set[str],
    *,
    recursive: bool,
    auto_group: bool,
) -> None:
    """Recursively walk directory tree for git repos."""
    if not current.is_dir():
        return

    try:
        entries = sorted(current.iterdir())
    except PermissionError:
        return

    for entry in entries:
        if not entry.is_dir():
            continue

        ## skip hidden dirs (but not .git itself — we check parent)
        if entry.name.startswith("."):
            continue

        if is_git_repo(entry):
            resolved = str(entry.resolve())
            if resolved not in seen:
                seen.add(resolved)
                group = None
                if auto_group and entry.parent != base:
                    group = entry.parent.name
                records.append(RepoRecord(
                    name=entry.name,
                    path=resolved,
                    group=group,
                ))
            ## don't recurse into a git repo
            continue

        if recursive:
            _walk(entry, base, records, seen, recursive=recursive, auto_group=auto_group)

File: internal/discover/discover.go
// Package discover walks a directory tree to find git repositories.
package discover

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/handy-sun/githand/internal/config"
	"github.com/handy-sun/githand/internal/git"
)

// Discover walks dir and returns all git repositories found.
// If recursive is true, it descends into subdirectories.
// If autoGroup is true, the immediate subdirectory name becomes the group.
func Discover(dir string, recursive, autoGroup bool) ([]config.Repo, error) {
	var repos []config.Repo
	seen := make(map[string]bool)

	err := walkDir(dir, dir, recursive, autoGroup, seen, &repos)
	if err != nil {
		return nil, err
	}
	return repos, nil
}

func walkDir(root, dir string, recursive, autoGroup bool, seen map[string]bool, repos *[]config.Repo) error {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil // skip unreadable dirs
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}

		fullPath := filepath.Join(dir, entry.Name())

		// skip hidden dirs (except when dir itself starts with .)
		if strings.HasPrefix(entry.Name(), ".") {
			// but check if it's a .git inside a non-git dir — skip
			continue
		}

		// check if this directory is a git repo
		if git.IsRepo(fullPath) {
			topLevel, err := git.TopLevel(fullPath)
			if err != nil {
				continue
			}

			if seen[topLevel] {
				continue
			}
			seen[topLevel] = true

			name := filepath.Base(topLevel)
			group := ""
			if autoGroup {
				group = computeGroup(root, topLevel)
			}

			*repos = append(*repos, config.Repo{
				Name:  name,
				Path:  topLevel,
				Group: group,
			})
			// don't recurse into a git repo's subdirectories
			continue
		}

		// not a git repo — recurse if allowed
		if recursive {
			_ = walkDir(root, fullPath, recursive, autoGroup, seen, repos)
		}
	}
	return nil
}

// computeGroup returns the subdirectory name under root as the group.
// e.g. root=/Users/qi/work, path=/Users/qi/work/nix/expnix -> "nix"
func computeGroup(root, path string) string {
	// resolve symlinks to handle macOS /var -> /private/var
	root, _ = filepath.EvalSymlinks(root)
	path, _ = filepath.EvalSymlinks(path)
	rel, err := filepath.Rel(root, path)
	if err != nil {
		return ""
	}
	parts := strings.Split(filepath.ToSlash(rel), "/")
	if len(parts) > 1 {
		return parts[0]
	}
	return ""
}
